### ISSUE TITLE
Implement a layout engine

### ISSUE THREAD

**Issue by yaycmyk** [tmpvar/jsdom#1322]
I was taking a look at https://github.com/facebook/css-layout and it seems fairly complete.

A layout engine is I think the last major piece of the puzzle that JSDOM needs. I know it's a bit of an undertaking...

**Comment by domenic** [comment]
It only does flexbox though, not e.g. relative or absolute or static positioning.

**Comment by Sebmaster** [comment]
css-layout only implements flexbox. It'd not even be possible to parse only newly developed sites people are trying to parse with jsdom. We talked about implementing a layouting engine at jsconf.eu - the will is there, but priority definitely isn't.

The problem is, when you start going down that rabbit hole there's a lot to consider - we'd have to include font rendering (or at least font width measurement) to get even the simplest websites to work. All those things could be a bit reduced (for example only emulating a monospace font), but they're nontheless there and there won't be any correct results without taking all that into consideration.

**Comment by inikulin** [comment]
Can't it be accomplished with native deps like Skia? It's not the best option, but it can be made as an optional extension (like node-canvas at the moment).

**Comment by yaycmyk** [comment]
@domenic It says on the README they support relative and static. The absence of the other `display` types is annoying though

**Comment by yaycmyk** [comment]
There's also this one: https://github.com/tategakibunko/nehan.js, though the docs are a little cryptic.

**Comment by Sebmaster** [comment]
nehan.js:
> Available on IE8+, Firefox3.5+, Google Chrome4+, Safari3+, Opera10+ etc.

Makes me think they're using dom methods to render that stuff, which makes it useless to implement rendering.

>  absolute, relative and static

That's the simple ones, but doesn't take into account overflows, etc. It basically only supports boxes and doesn't take into account their contents (which are a big deal on the web though).

**Comment by yaycmyk** [comment]
@Sebmaster I haven't looked too deep into it, but IE8+ could be due to using certain ES3-5+ things that the older browsers didn't support.

**Comment by inikulin** [comment]
Well, implementing layout engine seems like an interesting hobby task for years =) So, I decided to give it a try when things will settle down a bit with my primary projects. Meanwhile, I've decided to improve my knowledge on internals by reading CSS spec (interesting reading BTW) and made some basic research. Regarding font measurement: seems like we can do it already without native code.
We have [opentype.js](https://github.com/nodebox/opentype.js) for fonts parsing. It can [give us SVG path](https://github.com/nodebox/opentype.js#fontgetpathtext-x-y-fontsize-options) for the given text with the given font. From path we can get text BBox (example implementation in Raphael: https://github.com/DmitryBaranovskiy/raphael/blob/master/dev/raphael.core.js#L1592)

**Comment by inikulin** [comment]
On the other hand, for layouting we will need somekind of threading, since it's a lot of computation. But still there is nothing like that in JS. As far as I remember, I've seen some discussion regarding some kind of threading model in TC39 notes, but seems like it take years to take it to the actual implementation (I believe @domenic can tell more). The sad part is that we don't have any reliable alternatives.

**Comment by inikulin** [comment]
P.s. merry xmas :santa:

**Comment by Joris-van-der-Wel** [comment]
WebWorker + Transferable objects is probably the best bet for now. jsdom also runs in other browsers so any node specific feature would not be appropriate.

**Comment by inikulin** [comment]
We don't have WebWorkers in Node. It will require native extension.

**Comment by Joris-van-der-Wel** [comment]
I know, however we can extend node ourselves. Extending all the browsers is harder.

Being able to render to canvas would be nifty, browser inception

**Comment by guidobouman** [comment]
The layout engine is not going to happen anytime soon, or is it?

**Comment by Sebmaster** [comment]
> The layout engine is not going to happen anytime soon, or is it?

No.

**Comment by guidobouman** [comment]
How about https://github.com/trevorlinton/webkit.js? :trollface:

On a bit more serious note: was work on this ever started? Or is this something jsdom is just not designed to do, and should we get Headless Chrome instead?

**Comment by Sebmaster** [comment]
> On a bit more serious note: was work on this ever started?

Not really, I think the current state of affairs is that it'd be way to slow to actually productively use, although it's never been tried and would make for interesting results in any case.

**Comment by domenic** [comment]
I think it would be roughly a full-time job for 3-6 months for a very talented engineer. But, I think it could be done, and the result would probably be pretty fast actually.

But yeah, realistically this is probably not going to happen :)

**Comment by guidobouman** [comment]
Just acting really naive here, but what if this was sponsored by browser vendors? Would this give jsdom advantages over Headless Chrome + Puppeteer? I think it would, right? As the bridge would be a lot smaller.

**Comment by nicoburns** [comment]
If WASM is acceptable then https://github.com/DioxusLabs/taffy (disclosure: I maintain this library) implements Flexbox, CSS Grid, and is about to land Block layout. However inline (so text), inline-block, and float layout would still need to be implemented on top of this.
